src/OpenLoco/src/Map/TileManager.cpp

```diff
@@ -1043,6 +1063,210 @@ namespace OpenLoco::World::TileManager
+        if (surface->var_4_E0() > 0)
+        {
+            surface->setVar4SLR5(0);
+
+            Ui::ViewportManager::invalidate(pos, surface->baseHeight(), surface->baseHeight() + 32, ZoomLevel::eighth);
+        }
+    }
+
+    // 0x00468651
+    uint32_t adjustSurfaceHeight(World::Pos2 pos, SmallZ targetBaseZ, uint8_t slopeFlags, std::set<World::Pos3, LessThanPos3>& removedBuildings, uint8_t flags)
+    {
+        if (!validCoords(pos))
+        {
+            GameCommands::setErrorText(StringIds::off_edge_of_map);
+            return GameCommands::FAILURE;
+        }
+
+        if (targetBaseZ < 4)
+        {
+            GameCommands::setErrorText(StringIds::error_too_low);
+            return GameCommands::FAILURE;
+        }
+        if (targetBaseZ > 160
+            || (targetBaseZ == 160 && (slopeFlags & 0x1F) != 0)
+            || (targetBaseZ == 156 && (slopeFlags & 0x10) != 0))
+        {
+            GameCommands::setErrorText(StringIds::error_too_high);
+            return GameCommands::FAILURE;
+        }
+
+        currency32_t totalCost = 0;
+
+        if (flags & GameCommands::Flags::apply)
+        {
+            removeSurfaceIndustry(pos);
+
+            if (!isEditorMode())
+            {
+                setTerrainStyleAsCleared(pos);
+            }
+
+            auto clearHeight = getHeight(pos).landHeight;
+            removeAllWallsOnTile(toTileSpace(pos), clearHeight / 4);
+        }
+
+        // Compute cost of landscape operation
+        auto tile = get(pos);
+        auto* surface = tile.surface();
+        auto landObj = ObjectManager::get<LandObject>(surface->terrain());
+        totalCost += Economy::getInflationAdjustedCost(landObj->costFactor, landObj->costIndex, 10);
+
+        // If surface is in use as a water(route), and has water, ensure we don't remove it
+        if (surface->hasType6Flag() && surface->water() > 0)
+        {
+            auto waterHeight = (surface->water() - 1) * kMicroToSmallZStep;
+            auto testHeight = targetBaseZ;
+            if (slopeFlags & 0x1F)
+            {
+                testHeight += kSmallZStep;
+                if (slopeFlags & SurfaceSlope::doubleHeight)
+                {
+                    testHeight += kSmallZStep;
+                }
+            }
+
+            if (waterHeight < testHeight)
+            {
+                GameCommands::setErrorText(StringIds::water_channel_currently_needed_by_ships);
+                return GameCommands::FAILURE;
+            }
+        }
+
+        // Bind our local vars to the tile clear function
+        auto clearFunc = [pos, &removedBuildings, flags, &totalCost](TileElement& el) {
+            return TileClearance::clearWithDefaultCollision(el, pos, removedBuildings, flags, totalCost);
+        };
+
+        QuarterTile qt(0xF, 0);
+        auto testHeight = TileManager::getSurfaceCornerHeight(surface);
+        if (!TileClearance::applyClearAtAllHeights(pos, testHeight / 4, testHeight / 4, qt, clearFunc))
```

Comment:
I think this is wrong. `testHeight` is already in `smallZ`, so we shouldn't need to divide it further.